telemetry/src/lib.rs

```diff
@@ -0,0 +1,235 @@
+use std::{collections::VecDeque, sync, sync::Arc, time::Duration};
+
+use observability_deps::tracing::error;
+use serde::Serialize;
+use tokio::sync::mpsc;
+
+/// This is the core type exposed from this crate which can be
+/// cloned in different places to send the `InternalTelemetryMessage`.
+#[derive(Clone)]
+pub struct TelemetryHandle {
+    /// mpsc sender, safe to clone
+    sender: mpsc::Sender<InternalTelemetryMessage>,
+}
+
+impl TelemetryHandle {
+    pub async fn new(
+        instance_id: String,
+        os: String,
+        influx_version: String,
+        storage_type: String,
+        cores: u32,
+    ) -> Self {
+        let store = TelemetryStore::new(instance_id, os, influx_version, storage_type, cores);
+        // TODO: decide the buffer size
+        let (sender, receiver) = mpsc::channel(100);
```

Comment:
When I was at Fastly rewriting our entire metrics pipeline we had a bound of about 10k given the amount of reqs per second (thousands easily). It ended up being faster to just use a lock on the map to the stats and let the handle be able to update the stats rather than even using a channel.

If we're updating these after every single write for instance I fear the queue will fill up really quick and we'll start applying back pressure where we really don't want too, regardless of the queue size.

Reply:
Yea - I've hard coded it for now

Reply:
@mgattozzi - thanks, I was hoping the metrics gathering will get out of the critical path for any writes/reads when following the message passing model. If you think that having lock on the store directly will be quick enough anyway, happy to change the implementation to expose the `TelemetryStore` directly instead of going through the `TelemetryHandle`.